ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/anomalylocalization/AnomalyLocalizerImpl.java

```diff
@@ -0,0 +1,381 @@
+        Output output = new Output();
+        input.getAggregations().stream().forEach(agg -> localizeByBuckets(input, agg, output, notifyOnce(listener)));
+    }
+
+    private void localizeByBuckets(Input input, AggregationBuilder agg, Output output, ActionListener<Output> listener) {
+        LocalizationTimeBuckets timeBuckets = getTimeBuckets(input);
+        getOverallAggregates(input, timeBuckets, agg, output, listener);
+    }
+
+    private void getOverallAggregates(Input input, LocalizationTimeBuckets timeBuckets, AggregationBuilder agg, Output output,
+                                      ActionListener<Output> listener) {
+        MultiSearchRequest searchRequest = newSearchRequestForOverallAggregates(input, agg, timeBuckets);
+        client.multiSearch(searchRequest, wrap(r -> onOverallAggregatesResponse(r, input, agg, output, timeBuckets, listener),
+                listener::onFailure));
+    }
+
+    private void onOverallAggregatesResponse(MultiSearchResponse response, Input input, AggregationBuilder agg, Output output,
+                                             LocalizationTimeBuckets timeBuckets, ActionListener<Output> listener) {
+        Output.Result result = new Output.Result();
+        List<Map.Entry<Long, Long>> intervals = timeBuckets.getAllIntervals();
+        for (int i = 0; i < intervals.size(); i++) {
+            double value = getDoubleValue((SingleValue) response.getResponses()[i].getResponse().getAggregations().get(agg.getName()));
+
+            Output.Bucket bucket = new Output.Bucket();
+            bucket.setStartTime(intervals.get(i).getKey());
+            bucket.setEndTime(intervals.get(i).getValue());
+            bucket.setOverallAggValue(value);
+            result.getBuckets().add(bucket);
+        }
+
+        output.getResults().put(agg.getName(), result);
+        getLocalizedEntities(input, agg, result, output, listener);
+    }
+
+    private void getLocalizedEntities(Input input, AggregationBuilder agg, Output.Result result, Output output,
+                                      ActionListener<Output> listener) {
+        if (setBase(result, input)) {
+            Counter counter = new HybridCounter();
+            result.getBuckets().stream().filter(e -> e.getBase().isPresent() && e.getBase().get().equals(e))
+                    .forEach(e -> processBaseEntry(input, agg, result, e, counter, Optional.empty(), output, listener));
+        }
+        outputIfResultsAreComplete(output, listener);
+    }
+
+    private void outputIfResultsAreComplete(Output output, ActionListener<Output> listener) {
+        if (output.getResults().values().stream().allMatch(this::isResultComplete)) {
+            listener.onResponse(output);
+        }
+    }
+
+    private boolean isResultComplete(Output.Result result) {
+        return result.getBuckets().stream().allMatch(e -> e.getCompleted() == null || e.getCompleted().get() == true);
```

Comment:
curiosity question: why we consider the result is complete when `getCompleted()` filed is null?

Reply:
added a comment that when a bucket does not need localization, such as the base bucket or the bucket has no change over the base bucket.